vpn/client.go

```diff
@@ -79,16 +90,25 @@ func NewClientFromOptions(opt *Options) *Client {
 	if opt == nil {
 		return &Client{}
 	}
-	if opt.Log != nil {
-		logger = opt.Log
-	}
 	return &Client{
 		Opts:    opt,
 		tunInfo: &tunnelInfo{},
 		Dialer:  &net.Dialer{},
 	}
 }
 
+//
+// observability
+//
+
+// emit sends the passed stage into any configured EventListener
+func (c *Client) emit(stage uint16) {
+	if c.EventListener == nil {
+		return
+	}
+	c.EventListener <- stage
```

Comment:
How about

```suggestion
	select {
	case c.EventListener <-stage:
	default:
		// don't deliver
	}
```

This code is robust to a `nil` channel because [sending on a nil channel blocks forever](https://golangbyexample.com/send-receive-nil-channel-go/) but the second arm guarantees that we don't ever block on the channel.

If you go down this road, you can document that an interested user should create a sufficiently buffered channel, such that you can actually be ~sure you are not losing messages if you're unlucky and the sender sends when the receiver is not receiving (i.e., a 10 or 20-sized buffer would make losing messages super unlikely, given the relative rates at which you are writing and reading the channels)

Reply:
changed to select, and added documentation to EventListener.